Fix trailing newline in JUnit formatter

ilammy opened the pull request:
`write_message()` expects messages to contain no newlines.

Fixes https://github.com/rust-lang/rust/issues/149436

ilammy committed:
rustdoc: Write newline differently

Fix the panic in write_message() which expects messages to contain
no embedded newlines. We still want a trailing newline at the end
of the file though, so write it in different manner.

Doctest runner no longer panics, but the output is kinda broken
when `compile_fail` doctests are present. This is because they
are not mergeable.

chenyukang commented:
emm, seems there is no unit tests for all the formatters (not the issue for this PR).

ilammy committed:
rustdoc: Test --format=junit

Copied validate_junit.py from libtest-junit.

JUnit format works well in edition 2021, but is currently broken
in edition 2024 by the mergeable doctest report.

ilammy committed:
libtest: Stricter XML validation

Expect the entire input to be a single XML document, instead of
reading it line by line. This detects trailing junk better.

ilammy commented:
Whoops. I should have caught onto that when my `./x test` passed only with `--stage 2`. For some reason I thought the big pipeline handles it implicitly. Turns out, `rustdoc` is a bit special in that stage 1 builds for some reason I don't fully understand use stage 0's libtest, so any fix there are not propagated yet. Stage 2 is good though.

I'll go add `ignore-stage1` on the new test. I dunno how many existing ones technically need it as well.

ilammy committed:
rustdoc: Test only in stage2

stage1's rustdoc is using stage0's libtest which does not have
a fix from 069cf9dfc90, making the test run fail. Ensure that
this test is executed with everything recompiled in stage2.
